golang/cosmos/x/swingset/keeper/snapshotter.go

```diff
@@ -64,10 +65,19 @@ type SwingStoreExporter interface {
 }
 
 type SwingsetSnapshotter struct {
-	app            *baseapp.BaseApp
-	logger         log.Logger
-	exporter       SwingStoreExporter
-	blockingSend   func(action vm.Jsonable) (string, error)
+	sync.RWMutex      // must hold to read/write activeSnapshot
+	isConfigured      func() bool
+	takeSnapshot      func(height int64)
+	newRestoreContext func(height int64) sdk.Context
+	logger            log.Logger
+	exporter          SwingStoreExporter
+	blockingSend      func(action vm.Jsonable) (string, error)
+
+	// activeSnapshot points to the active snapshot.
+	// Must hold the snapshotter's RWMutex to read or write it.
+	// The main goroutine moves it from nil to non-nil,
+	// and the goroutine spawned by InitiateSnapshot() moves it from
+	// non-nil back to nil.
 	activeSnapshot *activeSnapshot
```

Comment:
I carefully examined the locking code, and while I struggled to stay awake looking at it, I *believe* it made sense. I find myself wishing for an abstraction like 
`activeSnapshot "sync/atomic".Pointer[activeSnapshot]`, especially since the RWLock is only there to protect loads and stores of the `activeSnapshot` pointer.

Since we're already on Go 1.19, I wouldn't fight against using generics.  Consider it?

Reply:
Totally agree. [sync.atomic.XXXPointer](https://pkg.go.dev/sync/atomic) is the way to go.

Reply:
I took a look and the `"sync/atomic" package right away echoes the advice at the top of the [Go memory model](https://go.dev/ref/mem): if you're using this, you're probably doing something wrong. It then repeats the Go slogan "Share memory by communicating; don't communicate by sharing memory."

Inspired by this I tried using a more idiomatic "done" channel for the worker goroutine to signal that it had finished, letting the main thread be in sole control of updates to the `activeSnapshot` field. It actually leads to simplifications in the code and allows more expressive tests. PTAL.